File: libs/cua-driver/examples/jev-use/typescript/keepalive.ts
/**
 * Keep-alive (persistent connection) client for the daemon's JSONL socket
 * protocol — a fork research prototype, NOT an upstream proposal.
 *
 * The real daemon speaks one-request-per-connection: every RPC pays a full
 * socket connect + handshake + teardown. This client opens ONE connection
 * and sends every request down it, reading one response line per request.
 * Requests are serialized (the daemon answers in order), matching how the
 * synchronous recipe loop issues calls.
 *
 * What it is NOT: a protocol change to the daemon. The daemon side is
 * `mock_daemon.mjs --keep-alive`. Upstream, this would require the
 * auth-lifetime contract question in dq-3383 (authenticate once per
 * connection vs per request) to be answered first — see REPORT-keepalive.md.
 */
import net from "node:net";
import type { CallFn } from "./batched.js";

const CALL_TIMEOUT_MS = 10_000;

export class PersistentDaemonClient {
  private readonly socketPath: string;
  private connection: net.Socket | null = null;
  private buffer = "";
  private waiters: Array<{
    resolve: (value: Record<string, any>) => void;
    reject: (error: Error) => void;
  }> = [];
  private tail: Promise<void> = Promise.resolve();
  private closed = false;

  constructor(socketPath: string) {
    this.socketPath = socketPath;
  }

  private ensureConnected(): Promise<net.Socket> {
    if (this.connection) return Promise.resolve(this.connection);
    if (this.closed) return Promise.reject(new Error("client is closed"));
    return new Promise((resolve, reject) => {
      const connection = net.createConnection(this.socketPath);
      connection.setEncoding("utf8");
      const timer = setTimeout(() => {
        connection.destroy();
        reject(new Error("daemon connect timeout"));
      }, CALL_TIMEOUT_MS);
      connection.on("connect", () => {
        clearTimeout(timer);
        this.connection = connection;
        resolve(connection);
      });
      connection.on("data", (chunk: string) => this.onData(chunk));
      connection.on("error", (error) => this.onFatal(error));
      connection.on("close", () =>
        this.onFatal(new Error("daemon closed the connection")),
      );
    });
  }

  private onData(chunk: string): void {
    this.buffer += chunk;
    for (;;) {
      const newline = this.buffer.indexOf("\n");
      if (newline < 0) return;
      const line = this.buffer.slice(0, newline);
      this.buffer = this.buffer.slice(newline + 1);
      const waiter = this.waiters.shift();
      if (!waiter) continue; // stray line; keep the buffer moving
      try {
        waiter.resolve(JSON.parse(line));
      } catch (error) {
        waiter.reject(
          error instanceof Error ? error : new Error(String(error)),
        );
      }
    }
  }

  private onFatal(error: Error): void {
    this.connection = null;
    this.buffer = "";
    const waiters = this.waiters;
    this.waiters = [];
    for (const waiter of waiters) waiter.reject(error);
  }

  /** One serialized call over the persistent connection. */
  call(name: string, args: Record<string, unknown>): Promise<Record<string, any>> {
    const run = this.tail.then(async () => {
      const connection = await this.ensureConnected();
      return new Promise<Record<string, any>>((resolve, reject) => {
        const timer = setTimeout(() => {
          this.waiters = this.waiters.filter((w) => w.resolve !== resolve);
          reject(new Error(`daemon call timeout: ${name}`));
        }, CALL_TIMEOUT_MS);
        this.waiters.push({
          resolve: (value) => {
            clearTimeout(timer);
            resolve(value);
          },
          reject: (error) => {
            clearTimeout(timer);
            reject(error);
          },
        });
        connection.write(`${JSON.stringify({ method: "call", name, args })}\n`);
      });
    });
    // Keep the chain alive even if this call fails; next call reconnects.
    this.tail = run.then(
      () => undefined,
      () => undefined,
    );
    return run;
  }

  /** The CallFn shape the observation-path code already consumes (raw wire). */
  asCallFn(): CallFn {
    return (name, args) => this.call(name, args);
  }

  /**
   * Bench-convention CallFn: same unwrapping as the one-shot bench clients —
   * throws on !ok, returns result.structuredContent. Use this when comparing
   * arms, so both arms see identical payloads.
   */
  asUnwrappedCallFn(): (name: string, args: Record<string, unknown>) => Promise<unknown> {
    return async (name, args) => {
      const wire = (await this.call(name, args)) as {
        ok: boolean;
        error?: string;
        result?: { structuredContent: { error?: string } & Record<string, unknown> };
      };
      if (!wire.ok) throw new Error(`daemon error on ${name}: ${wire.error}`);
      const sc = wire.result?.structuredContent;
      if (sc && typeof sc.error === "string") {
        throw new Error(`daemon error on ${name}: ${sc.error}`);
      }
      return sc;
    };
  }

  async close(): Promise<void> {
    this.closed = true;
    const connection = this.connection;
    this.connection = null;
    if (connection) {
      await new Promise<void>((resolve) => {
        connection.on("close", () => resolve());
        connection.end();
        setTimeout(resolve, 1000).unref?.();
      });
    }
  }
}

File: libs/cua-driver/examples/jev-use/python/keepalive.py
"""Keep-alive (persistent connection) client for the daemon's JSONL socket
protocol — a fork research prototype, NOT an upstream proposal.

The mock daemon speaks one-request-per-connection: every RPC pays a full
socket connect + handshake + teardown. This client opens ONE unix-socket
connection and sends every request down it, reading one response line per
request. Requests are serialized (the daemon answers in order), matching
how the synchronous recipe loop issues calls.

What it is NOT: a protocol change to the daemon. The daemon side is
``mock_daemon.py --keep-alive``. Upstream, this would require the
auth-lifetime contract question in dq-3383 (authenticate once per
connection vs per request) to be answered first.

Note on the Python run.py loop: it drives the real driver through MCP
stdio (``Driver`` in run.py), which is already a persistent connection —
there is no per-call connect cost there. This prototype is for the JSONL
unix-socket protocol path (the TS loop's mock_daemon.mjs world), answering:
if the Python loop ever spoke that protocol, how large is the per-call
connection overhead, and does a persistent client recover it?

Mirrors typescript/keepalive.ts on the fork's muse/keepalive-prototype
branch.
"""

from __future__ import annotations

import asyncio
import json
from collections import deque
from typing import Any, Awaitable, Callable, Mapping

CALL_TIMEOUT_S = 10.0


class DaemonCallError(RuntimeError):
    """Raised when a single daemon RPC fails (transport, timeout, or error payload)."""


class PersistentDaemonClient:
    """One unix-socket connection reused for every daemon call.

    Calls are serialized: each request is sent in turn and matched to the
    next response line, which is what the daemon's in-order answering
    guarantees. If the connection dies, pending waiters are rejected and
    the next call reconnects transparently.
    """

    def __init__(self, socket_path: str) -> None:
        self._socket_path = socket_path
        self._reader: asyncio.StreamReader | None = None
        self._writer: asyncio.StreamWriter | None = None
        self._waiters: deque[asyncio.Future] = deque()
        self._read_task: asyncio.Task | None = None
        self._send_lock = asyncio.Lock()
        self._closed = False

    async def _ensure_connected(self) -> tuple[asyncio.StreamReader, asyncio.StreamWriter]:
        if self._writer is not None and not self._writer.is_closing():
            return self._reader, self._writer
        if self._closed:
            raise DaemonCallError("client is closed")
        try:
            reader, writer = await asyncio.wait_for(
                asyncio.open_unix_connection(self._socket_path), timeout=CALL_TIMEOUT_S
            )
        except (OSError, asyncio.TimeoutError) as exc:
            raise DaemonCallError(f"daemon connect failed: {exc}") from exc
        self._reader, self._writer = reader, writer
        self._read_task = asyncio.ensure_future(self._read_loop())
        return reader, writer

    async def _read_loop(self) -> None:
        try:
            while True:
                line = await self._reader.readline()
                if not line:
                    self._on_fatal(DaemonCallError("daemon closed the connection"))
                    return
                try:
                    waiter = self._waiters.popleft()
                except IndexError:
                    continue  # stray line; keep the loop moving
                if waiter.done():
                    continue
                try:
                    waiter.set_result(json.loads(line.decode("utf-8")))
                except Exception as exc:  # noqa: BLE001 - must resolve the waiter
                    waiter.set_exception(DaemonCallError(f"bad response line: {exc}"))
        except asyncio.CancelledError:
            raise
        except Exception as exc:  # noqa: BLE001 - connection-level failure
            self._on_fatal(exc)

    def _on_fatal(self, exc: BaseException) -> None:
        writer, self._writer = self._writer, None
        self._reader = None
        waiters, self._waiters = self._waiters, deque()
        if writer is not None:
            writer.close()
        if self._read_task is not None:
            self._read_task.cancel()
            self._read_task = None
        for waiter in waiters:
            if not waiter.done():
                if isinstance(exc, BaseException):
                    waiter.set_exception(exc)
                else:
                    waiter.set_exception(DaemonCallError(str(exc)))

    async def call(self, name: str, args: Mapping[str, Any]) -> Mapping[str, Any]:
        """One serialized call over the persistent connection."""
        async with self._send_lock:
            _, writer = await self._ensure_connected()
            loop = asyncio.get_running_loop()
            waiter: asyncio.Future = loop.create_future()
            self._waiters.append(waiter)
            writer.write(json.dumps({"method": "call", "name": name, "args": dict(args)}).encode("utf-8") + b"\n")
            try:
                await writer.drain()
            except (ConnectionError, asyncio.IncompleteReadError) as exc:
                self._on_fatal(exc)
                raise DaemonCallError(f"daemon write failed on {name}: {exc}") from exc
            try:
                return await asyncio.wait_for(asyncio.shield(waiter), timeout=CALL_TIMEOUT_S)
            except asyncio.TimeoutError as exc:
                try:
                    self._waiters.remove(waiter)
                except ValueError:
                    pass
                if not waiter.done():
                    waiter.cancel()
                raise DaemonCallError(f"daemon call timeout: {name}") from exc

    def as_call_fn(self) -> Callable[[str, Mapping[str, Any]], Awaitable[Mapping[str, Any]]]:
        """The raw-wire CallFn shape: returns the daemon's JSON envelope."""
        return self.call

    def as_unwrapped_call_fn(self) -> Callable[[str, Mapping[str, Any]], Awaitable[Any]]:
        """Bench-convention CallFn: throw on !ok, return structuredContent.

        Use when comparing arms so both arms see identical payloads.
        """

        async def unwrapped(name: str, args: Mapping[str, Any]) -> Any:
            wire = await self.call(name, args)
            if not isinstance(wire, dict) or not wire.get("ok"):
                error = wire.get("error") if isinstance(wire, dict) else None
                raise DaemonCallError(f"daemon error on {name}: {error}")
            structured = wire.get("result", {}).get("structuredContent")
            if isinstance(structured, dict) and isinstance(structured.get("error"), str):
                raise DaemonCallError(f"daemon error on {name}: {structured['error']}")
            return structured

        return unwrapped

    async def close(self) -> None:
        self._closed = True
        if self._read_task is not None:
            self._read_task.cancel()
            self._read_task = None
        writer, self._writer = self._writer, None
        self._reader = None
        if writer is not None:
            writer.close()
            try:
                await writer.wait_closed()
            except (ConnectionError, asyncio.TimeoutError):
                pass
